pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -61,12 +62,141 @@ type replicaCircuitBreaker struct {
 	stopper *stop.Stopper
 	r       replicaInCircuitBreaker
 	st      *cluster.Settings
+	cancels CancelStorage
 	wrapped *circuit.Breaker
+
+	versionIsActive int32 // atomic
+}
+
+// Register takes a cancelable context and its cancel function, and registers
+// them with the circuit breaker. If the breaker is already tripped, its error
+// is returned immediately and the caller should not continue processing the
+// request. Otherwise, the cancel function is invoked when the breaker trips.
+// The caller is provided with a token and signaller for use in a call to
+// Unregister upon request completion. That method also takes the error (if any)
+// resulting from the request to ensure that in the case of a tripped breaker,
+// the error reflects this fact.
+func (br *replicaCircuitBreaker) Register(
+	ctx context.Context, cancel func(),
+) (_token interface{}, _ signaller, _ error) {
+	brSig := br.Signal()
+
+	if brSig.C() == nil {
```

Comment:
> This nil behavior should be specified in the Breaker.Signal() doc comment.

Are you sure? This is an internal implementation detail. Note that the `nil` channel originates in `(*replicaCircuitBreaker).Signal` and is only acted upon in `Register`/`Unregister`. "External" callers shouldn't special-case on the nil channel (at least not past the point of noting that since this channel is nil, it's obviously never going to close).

>  I suppose that may be fine since we specify in the doc comment that the signaller must be passed back in to Unregister().

Yeah, this is specified, but I can make it nicer - the signaller can sit in the token, where the caller has no way to replace it. Adding a TODO since I can't quite get to it right now.

Reply:
> Yeah, this is specified, but I can make it nicer - the signaller can sit in the token, where the caller has no way to replace it. Adding a TODO since I can't quite get to it right now.

Eh, this would require more allocs or a second pool or breaking the CancelStorage abstraction. Going to leave as is.

Reply:
> Are you sure? This is an internal implementation detail.

Well, nothing's really stopping anyone from replacing/wrapping the returned signaller with something that removes `C()` before passing it back in, which would leak a signaller. That's probably fine here, since I doubt anyone's going to be doing anything too fancy outside of the breaker itself, but just thought it might be worth mentioning somewhere.